### PR TITLE
[Bugfix] Fixed bug where the symbol solver can't resolve types in classes that…

### PR DESCRIPTION
**Pull request by jjbr**
… have an annotation as ancestor. The symbol solver tries to get the type parameters of parents which previously failed with an `UnsupportedOperationException`. The pull request includes both test and fix. Relates to #1833.

### CODE REVIEW

**Comment by matozoid** [comment]
Thanks! Looks fine to me.

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/core/resolution/Context.java [hunk 1]
```diff
@@ -192,7 +192,7 @@
                 methodUsage = ((TypeVariableResolutionCapability) methodDeclaration)
                                       .resolveTypeVariables(this, argumentsTypes);
             } else {
-                throw new UnsupportedOperationException();
+                return Optional.empty();
```
Comment (ftomassetti) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/core/resolution/Context.java]: mmm, why a method declaration should not have the TypeVariableResolutionCapability? Is this ok?

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/reflectionmodel/ReflectionAnnotationDeclaration.java [hunk 2]
```diff
@@ -154,8 +151,8 @@ public String getName() {
 
     @Override
     public List<ResolvedTypeParameterDeclaration> getTypeParameters() {
-        // TODO #1840
-        throw new UnsupportedOperationException();
+        // Annotation declarations cannot have type parameters - i.e. we can always return an empty list.
+        return Collections.emptyList();
```
Comment (ftomassetti) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/reflectionmodel/ReflectionAnnotationDeclaration.java:160]: Good! Maybe we can move this to the javadoc?

**Review by ftomassetti — COMMENTED** [review]
I have just a doubt about the Optional.empty

**Comment by jjbr** [comment]
I've now added a couple of Javadoc comments and also fixed the corresponding methods in the other AnnotationDeclarations. The `UnsupportedOperationException` is back as well, that behaviour seems more desirable than to silently return an unresolved symbol, you're completely right!

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/core/resolution/Context.java [hunk 3]
```diff
@@ -192,7 +192,8 @@
                 methodUsage = ((TypeVariableResolutionCapability) methodDeclaration)
                                       .resolveTypeVariables(this, argumentsTypes);
             } else {
-                throw new UnsupportedOperationException();
+                throw new UnsupportedOperationException("Resolved method declarations should have the " +
+                                                        TypeVariableResolutionCapability.class.getName() + ".");
```
Comment (matozoid) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/core/resolution/Context.java:196]: Thanks so much for adding a message!

**Comment by ftomassetti** [comment]
Updating branch. When tests are passing I would consider this good to go

**Pull request merged**